### PR TITLE
 Add nil literal support

### LINKED ISSUE #22359: Add nil literal to the new parser
**Description:**
Refer: https://ballerina.io/spec/lang/2020R1/#nil-literal

### PR DESCRIPTION
**Pull request by lochana-chathura**
Fixes #22359 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -5403,4 +5414,28 @@ private STNode parseLocalTypeDefinitionStatement(STNode annots) {
...
+            case OPEN_PAREN_TOKEN:
+                startContext(ParserRuleContext.NIL_LITERAL);
+                STNode openParenthesisToken = parseOpenParenthesis();
+                STNode closeParenthesisToken = parseCloseParenthesis();
+                endContext();
+                return STNodeFactory.createNilLiteral(openParenthesisToken, closeParenthesisToken);
+            default:
+                Solution sol = recover(token, ParserRuleContext.NIL_LITERAL);
+                return sol.recoveredNode;
+        }
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:5439]: Is this approach ok?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/SyntaxKind.java [hunk 2]
```diff
@@ -184,6 +185,7 @@
     SERVICE_BODY(3023),
     ANNOTATION(3024),
     METADATA(3025),
+    NIL_LITERAL(3026),
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/SyntaxKind.java:188]: Couldn't place this under section where other literals are declared as `parserTestutils.java` has the following method.
```public static boolean isTerminalNode(SyntaxKind syntaxKind) {
        return SyntaxKind.IMPORT_DECLARATION.compareTo(syntaxKind) > 0 || syntaxKind == SyntaxKind.EOF_TOKEN;
    }```

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 3]
```diff
@@ -814,7 +817,8 @@ private Result seekMatch(ParserRuleContext currentCtx, int lookahead, int curren
                 case ACCESS_EXPRESSION:
                     return seekInAccessExpression(currentCtx, lookahead, currentDepth, matchingRulesCount);
                 case BASIC_LITERAL:
-                    hasMatch = isBasicLiteral(nextToken.kind);
+                    STToken nextNextToken = this.tokenReader.peek(lookahead + 1);
+                    hasMatch = isBasicLiteral(nextToken.kind, nextNextToken.kind);
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:821]: Don't need to handle `()` in basic literals.. It was added to deal with literals with a single token. So may be we could only handle `null` here.

For `()` lets use `NIL_LITERAL` and add it to the `EXPRESSIONS` alternatives array.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 4]
```diff
@@ -1358,10 +1367,14 @@ private ParserRuleContext getNextRule(ParserRuleContext currentCtx, int nextLook
...
-                endContext();
-                // After parsing nil type descriptor all the other parsing is same as next rule of simple type
-                return getNextRuleForTypeDescriptor();
-            }
+                    endContext();
+                    // After parsing nil type descriptor all the other parsing is same as next rule of simple type
+                    return getNextRuleForTypeDescriptor();
+                }
+                if (parentCtx == ParserRuleContext.NIL_LITERAL) {
+                    endContext();
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:1375]: Eventhough the `endContext()` is called here, `startContext()` is not called at the first switch-case of `getNextRule()` method.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 5]
```diff
@@ -1644,6 +1657,9 @@ private ParserRuleContext getNextRule(ParserRuleContext currentCtx, int nextLook
                 return ParserRuleContext.EXPRESSION_RHS;
             case LOCAL_TYPE_DEFINITION_STMT:
                 return ParserRuleContext.TYPE_KEYWORD;
+            case NULL_KEYWORD:
+                endContext(); // end nil-literal
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:1661]: A context is not started for null keyword.. So ending the context here is not correct. 

We might not need any context for null keyword at all. It is very similar to string literal, int literal etc. A context is needed when a token can occur in different contexts/places  and if we can't determine whats next rule should be.. Then we make use of the parent context.

File: compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json [hunk 6]
```diff
@@ -1425,6 +1425,21 @@
           "type": "Node"
         }
       ]
+    },
+    {
+      "name": "NilLiteral",
```
Comment (sameerajayasoma) [inline compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json:1430]: Use NilLiteralNode as per the new convention

File: compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json [hunk 7]
```diff
@@ -1425,6 +1425,21 @@
...
+      "base": "Node",
+      "kind": "NIL_LITERAL",
+      "attributes": [
+        {
+          "name": "firstToken",
```
Comment (sameerajayasoma) [inline compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json:1435]: Should this be openParenToken?

File: compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json [hunk 8]
```diff
@@ -1425,6 +1425,21 @@
...
+          "type": "Token"
+        },
+        {
+          "name": "secondToken",
```
Comment (sameerajayasoma) [inline compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json:1439]: Should this be closeParanToken

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 1 above)
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:5439]: @SupunS No switch case means, `( )` is considered to be `NIL_LITERAL` kind and `NULL` is considered to be `NULL_KEYWORD` kind, and will have separate `pareseNilLiteral( )`, `pareseNullKeyword( )` methods ?

Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:5439]: yes. For the null keyword, we can do the same way as, string literal

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java (hunk 3 above)
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java]: Adding `NIL_LITERAL` to `EXPRESSIONS` array cause 
```public function foo hello (
}``` to be generated as ```public function foo  MISSING[(]hello  MISSING[] MISSING[=](
 MISSING[] MISSING[)] MISSING[)] MISSING[{]}```
I will have a look on what's happening there.

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java]: Adding `NIL_LITERAL` to `EXPRESSIONS` array cause func definition with extra func name;  `public function foo hello (
}` to be generated as

`public function foo  MISSING[(]hello  MISSING[] MISSING[=](
 MISSING[] MISSING[)] MISSING[)] MISSING[{]}`

I will have a look on what's happening there

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/NilLiteralTest.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    public void testSimpleNilLiteral() {
+        test("()", "basic-literals/nil_literal_assert_01.json");
+        test("null", "basic-literals/nil_literal_assert_02.json");
+    }
+
+    @Test
+    public void testNilLiteral() {
+        test("() != null + a", "basic-literals/nil_literal_assert_03.json");
+    }
+}
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/NilLiteralTest.java:39]: Shall we add some more tests, to validate the usages of nil-literal (`()`), such as var-decl, defaultable-params, in object/record fields, etc?

**Comment by lochana-chathura** [comment]
Plz hold the PR. I'll add IsExpressionNode name change to this.

**Comment by lochana-chathura** [comment]
Change added!

**Pull request merged**